### ISSUE TITLE
'git clone' taking too long, occasional timeouts

### ISSUE THREAD

**Issue by pbrkr** [raspberrypi/firmware#175]
It took me several attempts to 'git clone' this repository as github kept timing out, the error I saw was:

> fatal: read error: Connection reset by peer
> fatal: early EOF
> fatal: index-pack failed

After several attempts one worked and the resulting bare repository has a size of 1.2GB. All I need are a specfic revision (0ac68cce44d4550c251172e8524100090e8211fa) of the files in boot, modules and opt/vc. Even the Hexxeh firmware repository (https://github.com/Hexxeh/rpi-firmare) is huge. With a sustained download speed of around 400KiB it's taking over an hour to clone the repository, much of which time is spent waiting for github to compress objects.

I'd argue that this is not a sustainable method of distributing firmware for the Raspberry Pi. The repository is a little over a year old and has under 200 revisions. How large is this going to be in another year? Or another 5?

I'd recommend that source files stay in a revision control system like this and binary firmware gets distributed as .tar.bz2 files (or equivalent, xz is very good these days) from an http/ftp server, the same way every other project distributes their binary files. I could provide a mirror like that myself, but this is to my knowledge the official source for firmware so my mirror would be a second-class citizen.

Basically, are there any alternative methods of fetching just the firmware I need on the Raspberry Pi itself, at a single revision which is not necessarily master, rather than cloning the entire repository? And if not, could you please consider providing such a service.

Many Thanks,

Paul Barker

**Comment by pbrkr** [comment]
@asb @popcornmix - please re-open this as it's now becoming very painful again. We need a way of automatically downloading a commit from this repository. Git shallow clones aren't useful as they only support downloading a branch or a tag, not an arbitrary commit. Artifacts are no use as the files are sometimes re-generated by GitHub which changes the checksums and our fetcher sees this as data corruption.

Git is not an appropriate tool for versioning binary files like the boot files. An alternative way to download the boot files would be much appreciated here.

**Comment by popcornmix** [comment]
What's wrong with downloading:
``https://github.com/raspberrypi/firmware/archive/<hash>.tar.gz`` ?

e.g. [this](https://github.com/raspberrypi/firmware/archive/0ac68cce44d4550c251172e8524100090e8211fa.tar.gz)

**Comment by pbrkr** [comment]
We use `md5sum` and `sha256sum` checks to protect the integrity of file downloads. GitHub re-generates those archives occasionally, resulting in different checksums. This causes downloads to fail for all downstream users of the meta-raspberrypi Yocto Project layer.

E.g. from a colleague earlier today:

```
WARNING: bcm2835-bootfiles-20190823-r3 do_fetch: Checksum mismatch for local file /sddm2/panto/freehold/build/downloads/7163480fff007dc98978899b556dcf06f8a462c8.tar.gz
Cleaning and trying again.
WARNING: bcm2835-bootfiles-20190823-r3 do_fetch: Renaming /sddm2/panto/freehold/build/downloads/7163480fff007dc98978899b556dcf06f8a462c8.tar.gz to /sddm2/panto/freehold/build/downloads/7163480fff007dc98978899b556dcf06f8a462c8.tar.gz_bad-checksum_ebf4f7a90d00440775c8388d0ee6f6b3
WARNING: bcm2835-bootfiles-20190823-r3 do_fetch: Checksum failure encountered with download of https://github.com/raspberrypi/firmware/archive/7163480fff007dc98978899b556dcf06f8a462c8.tar.gz - will attempt other sources if available
ERROR: bcm2835-bootfiles-20190823-r3 do_fetch: Fetcher failure for URL: 'https://github.com/raspberrypi/firmware/archive/7163480fff007dc98978899b556dcf06f8a462c8.tar.gz'. Checksum mismatch!
File: '/sddm2/panto/freehold/build/downloads/7163480fff007dc98978899b556dcf06f8a462c8.tar.gz' has md5 checksum ebf4f7a90d00440775c8388d0ee6f6b3 when b10044aa637ed170eef2ca6c1dcc8aba was expected
File: '/sddm2/panto/freehold/build/downloads/7163480fff007dc98978899b556dcf06f8a462c8.tar.gz' has sha256 checksum eeec3419f8e076e5dea327936a98873f4506b160d4d5d2b529dbb31dd74abcbc when 012bc542157d03d19c52dfb2ff9e828905d1991a8b33420f1a2e3730040c167f was expected
If this change is expected (e.g. you have upgraded to a new version without updating the checksums) then you can use these lines within the recipe:
SRC_URI[md5sum] = "ebf4f7a90d00440775c8388d0ee6f6b3"
SRC_URI[sha256sum] = "eeec3419f8e076e5dea327936a98873f4506b160d4d5d2b529dbb31dd74abcbc"
Otherwise you should retry the download and/or check with upstream to determine if the file has become corrupted or otherwise unexpectedly modified.

ERROR: bcm2835-bootfiles-20190823-r3 do_fetch: Fetcher failure for URL: 'https://github.com/raspberrypi/firmware/archive/7163480fff007dc98978899b556dcf06f8a462c8.tar.gz'. Unable to fetch URL from any source.
```
